Carbon.MassTransit/ReceiveEndpointObserver.cs

```diff
@@ -0,0 +1,62 @@
+﻿using Carbon.MassTransit;
+using MassTransit;
+using Microsoft.Extensions.Configuration;
+using System;
+using System.Threading.Tasks;
+
+namespace Carbon.MassTransit
+{
+    public class ReceiveEndpointObserver : IReceiveEndpointObserver
+    {
+        private readonly IConfiguration _configuration;
+        public ReceiveEndpointObserver(IConfiguration configuration)
+        {
+            _configuration = configuration;
+        }
+        public Task Completed(ReceiveEndpointCompleted completed)
+        {
+            return Task.FromResult(true);
+        }
+
+        public Task Faulted(ReceiveEndpointFaulted faulted)
+        {
+            //queue-type based exceptions handled here
+            if (faulted.Exception.InnerException is RabbitMQ.Client.Exceptions.OperationInterruptedException)
+            {
+                var fault = ((RabbitMQ.Client.Exceptions.OperationInterruptedException)faulted.Exception.InnerException);
+                ushort faultCode = fault.ShutdownReason.ReplyCode;
+                string faultMessage = fault.ShutdownReason.ReplyText;
+                //If it is an existing queue with different type, delete it and let MassTransit create again
+                if (faultCode == 406 && faultMessage.Contains("x-queue-type"))
+                {
+                    var massTransitSettings = _configuration.GetSection("MassTransit").Get<MassTransitSettings>();
+                    var busSettings = massTransitSettings.RabbitMq;
+                    var rabbitMqUql = $"amqp://{busSettings.Username}:{busSettings.Password}@{busSettings.Host}:{busSettings.Port}{busSettings.VirtualHost}";
+
+                    var factory = new RabbitMQ.Client.ConnectionFactory()
+                    {
+                        Uri = new Uri(rabbitMqUql)
+                    };
+                    var absolutePath = faulted.ReceiveEndpoint.InputAddress.AbsolutePath.Split('/');
+                    var queueName = absolutePath[absolutePath.Length - 1];
+                    using (var connection = factory.CreateConnection())
+                    {
+                        using (var channel = connection.CreateModel())
+                            channel.QueueDelete(queueName, false, false);
```

Comment:
Whats gonna happen if the queue is not empty? Isn't there a way to preserve messages?

Reply:
If it is not empty, queue will be purged and deleted. There is no way to wait until all the messages consumed and then delete. So they should handle with care while migrating.